Mathlib/RingTheory/Presentation.lean

```diff
@@ -0,0 +1,224 @@
+protected abbrev Quotient : Type (max w u) := P.Ring ⧸ P.ker
+
+/-- `P.Quotient` is `P.Ring`-isomorphic to `S` and in particular `R`-isomorphic to `S`. -/
+def quotientEquiv : P.Quotient ≃ₐ[P.Ring] S :=
+  Ideal.quotientKerAlgEquivOfRightInverse (f := Algebra.ofId P.Ring S) P.aeval_val_σ
+
+@[simp]
+lemma quotientEquiv_mk (p : P.Ring) : P.quotientEquiv p = algebraMap P.Ring S p :=
+  rfl
+
+@[simp]
+lemma quotientEquiv_symm (x : S) : P.quotientEquiv.symm x = P.σ x :=
+  rfl
+
+/--
+Dimension of a presentation defined as the cardinality of the generators
+minus the cardinality of the relations.
+
+Note: this definition is completely non-sensical for non-finite presentations and
+even then for this to make sense, you should assume that the presentation
+is a complete intersection.
+-/
+noncomputable def dimension : ℕ :=
+  Nat.card P.vars - Nat.card P.rels
+
+/-- A presentation is finite if there are only finitely-many
+relations and finitely-many relations. -/
+class IsFinite (P : Presentation.{t, w} R S) : Prop where
+  finite_vars : Finite P.vars
+  finite_rels : Finite P.rels
+
+attribute [instance] IsFinite.finite_vars IsFinite.finite_rels
+
+lemma ideal_fg_of_isFinite [P.IsFinite] : P.ker.FG := by
+  use (Set.finite_range P.relation).toFinset
+  simp [span_range_relation_eq_ker]
+
+/-- If a presentation is finite, the corresponding quotient is
+of finite presentation. -/
+instance [P.IsFinite] : FinitePresentation R P.Quotient :=
+  FinitePresentation.quotient P.ideal_fg_of_isFinite
+
+lemma finitePresentation_of_isFinite [P.IsFinite] :
+    FinitePresentation R S :=
+  FinitePresentation.equiv (P.quotientEquiv.restrictScalars R)
+
+section Construction
+
+section Localization
+
+variable (r : R) [IsLocalization.Away r S]
+
+open IsLocalization.Away
+
+private lemma ker_algebraMap_eq_span_range_relation_localizationAway :
+    RingHom.ker (aeval (S₁ := S) (Generators.localizationAway r).val) =
+      Ideal.span { C r * X () - 1 } := by
+  have : aeval (S₁ := S) (Generators.localizationAway r).val =
+      (mvPolynomialQuotientEquiv S r).toAlgHom.comp
+        (Ideal.Quotient.mkₐ R (Ideal.span {C r * X () - 1})) := by
+    ext x
+    simp only [Generators.localizationAway_vars, aeval_X, Generators.localizationAway_val,
+      AlgEquiv.toAlgHom_eq_coe, AlgHom.coe_comp, AlgHom.coe_coe, Ideal.Quotient.mkₐ_eq_mk,
+      Function.comp_apply]
+    rw [IsLocalization.Away.mvPolynomialQuotientEquiv_apply, aeval_X]
+  rw [this]
+  erw [← RingHom.comap_ker]
+  simp only [Generators.localizationAway_vars, AlgEquiv.toAlgHom_eq_coe, AlgHom.toRingHom_eq_coe,
+    AlgEquiv.toAlgHom_toRingHom]
+  show Ideal.comap _ (RingHom.ker (mvPolynomialQuotientEquiv S r)) =
+    Ideal.span {C r * X () - 1}
+  simp [RingHom.ker_equiv, ← RingHom.ker_eq_comap_bot]
+
+/-- If `S` is the localization of `R` away from `r`, we can construct a natural
+presentation of `S` as `R`-algebra with a single generator `X` and the relation `r * X - 1 = 0`. -/
+@[simps rels relation]
```

Comment:
Working with the `Generators` api a bit more, I've come to realize having simp lemmas on the indexing types are not a good thing. Maybe make `rels` `lemmasOnly`.